File: contracts/src/policy.rs
#![no_std]

use soroban_sdk::{
    contract, contractimpl, contractmeta, panic_with_error,
    token, Address, Env, String, symbol_short,
};

use crate::types::{DataKey, PaymentRecord, SpendingPolicy};

contractmeta!(
    key = "Description",
    val = "Proxima Spending Policy — programmable autonomous payment authorization for AI agents"
);

/// Approximate number of ledgers in 24 hours (5s per ledger)
const LEDGERS_PER_DAY: u32 = 17_280;

#[derive(Debug, Clone, Copy, PartialEq)]
#[repr(u32)]
pub enum PolicyError {
    PolicyNotFound      = 1,
    NotOwner            = 2,
    PolicyInactive      = 3,
    ExceedsPerTxLimit   = 4,
    ExceedsDailyLimit   = 5,
    InvalidAmount       = 6,
    UnauthorizedAgent   = 7,
    RecipientNotAllowed = 8,
}

impl soroban_sdk::TryFromVal<Env, soroban_sdk::Val> for PolicyError {
    type Error = soroban_sdk::ConversionError;
    fn try_from_val(_env: &Env, val: &soroban_sdk::Val) -> Result<Self, Self::Error> {
        let n: u32 = val.try_into()?;
        match n {
            1 => Ok(PolicyError::PolicyNotFound),
            2 => Ok(PolicyError::NotOwner),
            3 => Ok(PolicyError::PolicyInactive),
            4 => Ok(PolicyError::ExceedsPerTxLimit),
            5 => Ok(PolicyError::ExceedsDailyLimit),
            6 => Ok(PolicyError::InvalidAmount),
            7 => Ok(PolicyError::UnauthorizedAgent),
            8 => Ok(PolicyError::RecipientNotAllowed),
            _ => Err(soroban_sdk::ConversionError),
        }
    }
}

impl From<PolicyError> for soroban_sdk::Error {
    fn from(e: PolicyError) -> Self {
        soroban_sdk::Error::from_contract_error(e as u32)
    }
}

#[contract]
pub struct PolicyContract;

#[contractimpl]
impl PolicyContract {

    /// Create a new spending policy that authorizes an AI agent to make
    /// autonomous payments on behalf of the policy owner, within defined limits.
    ///
    /// # Arguments
    /// * `agent`              - Address of the agent being authorized
    /// * `max_per_tx`         - Maximum amount allowed per single transaction (stroops)
    /// * `daily_limit`        - Maximum total amount per day (stroops)
    /// * `asset`              - Asset code (e.g. "USDC")
    /// * `issuer`             - Asset issuer address
    /// * `allowed_recipient`  - Optional: restrict to a specific recipient
    ///
    /// Returns the new policy ID.
    pub fn create_policy(
        env: Env,
        agent: Address,
        max_per_tx: i128,
        daily_limit: i128,
        asset: String,
        issuer: Address,
        allowed_recipient: Option<Address>,
    ) -> u64 {
        let owner = env.current_contract_address();
        // Production: let owner: Address = env.invoker(); owner.require_auth();

        if max_per_tx <= 0 || daily_limit <= 0 {
            panic_with_error!(&env, PolicyError::InvalidAmount);
        }

        // Increment policy counter
        let policy_id: u64 = env
            .storage()
            .instance()
            .get(&DataKey::PolicyCount)
            .unwrap_or(0_u64)
            + 1;

        let policy = SpendingPolicy {
            id: policy_id,
            owner,
            agent: agent.clone(),
            max_per_tx,
            daily_limit,
            asset,
            issuer,
            spent_today: 0,
            last_reset_ledger: env.ledger().sequence(),
            total_spent: 0,
            is_active: true,
            created_at: env.ledger().sequence(),
            allowed_recipient,
        };

        let key = DataKey::Policy(policy_id);
        env.storage().persistent().set(&key, &policy);
        env.storage().persistent().extend_ttl(&key, 6_307_200, 6_307_200);
        env.storage().instance().set(&DataKey::PolicyCount, &policy_id);

        env.events().publish(
            (symbol_short!("policy"), symbol_short!("create")),
            policy_id,
        );

        policy_id
    }

    /// Execute an autonomous payment under a spending policy.
    ///
    /// This is the core function — it allows an authorized agent to pay
    /// a recipient WITHOUT requiring the owner to sign each transaction.
    /// The policy contract enforces all constraints on-chain.
    ///
    /// # Arguments
    /// * `policy_id`  - ID of the spending policy
    /// * `recipient`  - Address receiving the payment
    /// * `amount`     - Amount in stroops
    /// * `memo`       - Short description of the payment purpose
    ///
    /// Emits: `PaymentExecuted(policy_id, amount)`
    pub fn execute_payment(
        env: Env,
        policy_id: u64,
        recipient: Address,
        amount: i128,
        memo: String,
    ) -> PaymentRecord {
        let key = DataKey::Policy(policy_id);
        let mut policy: SpendingPolicy = env
            .storage()
            .persistent()
            .get(&key)
            .unwrap_or_else(|| panic_with_error!(&env, PolicyError::PolicyNotFound));

        // Only the authorized agent can call this
        policy.agent.require_auth();

        if !policy.is_active {
            panic_with_error!(&env, PolicyError::PolicyInactive);
        }

        if amount <= 0 {
            panic_with_error!(&env, PolicyError::InvalidAmount);
        }

        // Per-transaction limit check
        if amount > policy.max_per_tx {
            panic_with_error!(&env, PolicyError::ExceedsPerTxLimit);
        }

        // Optional recipient restriction
        if let Some(ref allowed) = policy.allowed_recipient {
            if *allowed != recipient {
                panic_with_error!(&env, PolicyError::RecipientNotAllowed);
            }
        }

        // Reset daily counter if 24h have passed
        let current_ledger = env.ledger().sequence();
        let ledgers_since_reset = current_ledger.saturating_sub(policy.last_reset_ledger);
        if ledgers_since_reset >= LEDGERS_PER_DAY {
            policy.spent_today = 0;
            policy.last_reset_ledger = current_ledger;
        }

        // Daily limit check
        if policy.spent_today + amount > policy.daily_limit {
            panic_with_error!(&env, PolicyError::ExceedsDailyLimit);
        }

        // Execute the actual token transfer via Stellar token interface
        let token_client = token::Client::new(&env, &policy.issuer);
        token_client.transfer(&env.current_contract_address(), &recipient, &amount);

        // Update policy state
        policy.spent_today += amount;
        policy.total_spent += amount;
        env.storage().persistent().set(&key, &policy);

        let record = PaymentRecord {
            policy_id,
            agent: policy.agent.clone(),
            recipient: recipient.clone(),
            amount,
            asset: policy.asset.clone(),
            memo,
            ledger: current_ledger,
        };

        env.events().publish(
            (symbol_short!("pay"), symbol_short!("exec")),
            (policy_id, amount),
        );

        record
    }

    /// Revoke a spending policy. Only the owner can revoke.
    pub fn revoke_policy(env: Env, policy_id: u64) {
        let key = DataKey::Policy(policy_id);
        let mut policy: SpendingPolicy = env
            .storage()
            .persistent()
            .get(&key)
            .unwrap_or_else(|| panic_with_error!(&env, PolicyError::PolicyNotFound));

        policy.owner.require_auth();
        policy.is_active = false;

        env.storage().persistent().set(&key, &policy);

        env.events().publish(
            (symbol_short!("policy"), symbol_short!("revoke")),
            policy_id,
        );
    }

    /// Get a spending policy by ID.
    pub fn get_policy(env: Env, policy_id: u64) -> SpendingPolicy {
        env.storage()
            .persistent()
            .get(&DataKey::Policy(policy_id))
            .unwrap_or_else(|| panic_with_error!(&env, PolicyError::PolicyNotFound))
    }

    /// Check if an agent is authorized under a given policy.
    pub fn is_authorized(env: Env, policy_id: u64, agent: Address) -> bool {
        if let Some(policy) = env
            .storage()
            .persistent()
            .get::<DataKey, SpendingPolicy>(&DataKey::Policy(policy_id))
        {
            policy.is_active && policy.agent == agent
        } else {
            false
        }
    }

    /// Return remaining daily spending allowance for a policy.
    pub fn remaining_allowance(env: Env, policy_id: u64) -> i128 {
        let policy: SpendingPolicy = env
            .storage()
            .persistent()
            .get(&DataKey::Policy(policy_id))
            .unwrap_or_else(|| panic_with_error!(&env, PolicyError::PolicyNotFound));

        let current_ledger = env.ledger().sequence();
        let ledgers_since_reset = current_ledger.saturating_sub(policy.last_reset_ledger);

        if ledgers_since_reset >= LEDGERS_PER_DAY {
            policy.daily_limit // full reset
        } else {
            policy.daily_limit - policy.spent_today
        }
    }

    /// Return total number of policies created.
    pub fn policy_count(env: Env) -> u64 {
        env.storage().instance().get(&DataKey::PolicyCount).unwrap_or(0)
    }
}

File: sdk/src/policy.ts
import {
  Contract,
  TransactionBuilder,
  BASE_FEE,
  nativeToScVal,
  scValToNative,
  xdr,
  Keypair,
  Transaction,
} from '@stellar/stellar-sdk';
import { rpc as SorobanRpc } from '@stellar/stellar-sdk';

import { ProximaError, ErrorCodes } from './types';
import type {
  SpendingPolicy,
  CreatePolicyParams,
  ExecutePaymentParams,
  PaymentRecord,
} from './types';
import {
  ResolvedConfig,
  createRpcServer,
  toStroops,
  fromStroops,
} from './stellar';

/**
 * PolicyClient — interact with the Proxima Spending Policy contract.
 *
 * This is the core of Proxima's autonomous payment capability.
 * It allows AI agents to make payments without per-transaction human approval,
 * within the boundaries defined by the policy owner.
 *
 * @example
 * ```ts
 * const policy = new PolicyClient(config)
 *
 * // Create a policy: authorize an agent to spend up to 10 USDC/day
 * const policyId = await policy.create({
 *   agent: 'GDATAAGENT...',
 *   maxPerTx: '0.50',
 *   dailyLimit: '10.00',
 *   asset: 'USDC',
 *   issuer: USDC_ISSUER.mainnet,
 * }, ownerKeypair)
 *
 * // Agent executes a payment autonomously (no owner signature needed)
 * await policy.executePayment({
 *   policyId,
 *   recipient: 'GRECIPIENT...',
 *   amount: '0.01',
 *   memo: 'API call #4821',
 * }, agentKeypair)
 * ```
 */
export class PolicyClient {
  private rpc: SorobanRpc.Server;
  private _contract: Contract | null = null;

  constructor(private config: ResolvedConfig) {
    this.rpc = createRpcServer(config);
  }

  /** Lazy contract accessor — validates the contract ID only when first used. */
  private get contract(): Contract {
    if (!this._contract) {
      if (!this.config.policyContractId) {
        throw new ProximaError(
          'Policy contract ID is not configured for this network.',
          ErrorCodes.CONTRACT_ERROR
        );
      }
      this._contract = new Contract(this.config.policyContractId);
    }
    return this._contract;
  }

  /**
   * Return total number of policies ever created on-chain.
   */
  async policyCount(): Promise<bigint> {
    const result = await this.rpc.simulateTransaction(
      this._buildSimulation('policy_count', [])
    );
    if (SorobanRpc.Api.isSimulationError(result)) return 0n;
    const raw = scValToNative(
      (result as SorobanRpc.Api.SimulateTransactionSuccessResponse).result!.retval
    );
    return BigInt(raw as number);
  }

  // ─── Read Methods ───────────────────────────────────────────────────────────

  /**
   * Retrieve a spending policy by ID.
   */
  async getPolicy(policyId: bigint): Promise<SpendingPolicy> {
    const result = await this.rpc.simulateTransaction(
      this._buildSimulation('get_policy', [nativeToScVal(policyId, { type: 'u64' })])
    );

    if (SorobanRpc.Api.isSimulationError(result)) {
      throw new ProximaError(
        `Policy ${policyId} not found`,
        ErrorCodes.POLICY_NOT_FOUND,
        result.error
      );
    }

    const raw = scValToNative((result as SorobanRpc.Api.SimulateTransactionSuccessResponse).result!.retval);
    return this._parsePolicy(raw);
  }

  /**
   * Check if an agent is authorized under a given policy.
   */
  async isAuthorized(policyId: bigint, agentAddress: string): Promise<boolean> {
    const result = await this.rpc.simulateTransaction(
      this._buildSimulation('is_authorized', [
        nativeToScVal(policyId, { type: 'u64' }),
        nativeToScVal(agentAddress, { type: 'address' }),
      ])
    );
    if (SorobanRpc.Api.isSimulationError(result)) return false;
    return scValToNative((result as SorobanRpc.Api.SimulateTransactionSuccessResponse).result!.retval) as boolean;
  }

  /**
   * Return remaining daily allowance for a policy (in stroops).
   */
  async remainingAllowance(policyId: bigint): Promise<bigint> {
    const result = await this.rpc.simulateTransaction(
      this._buildSimulation('remaining_allowance', [
        nativeToScVal(policyId, { type: 'u64' }),
      ])
    );
    if (SorobanRpc.Api.isSimulationError(result)) return 0n;
    const raw = scValToNative((result as SorobanRpc.Api.SimulateTransactionSuccessResponse).result!.retval);
    return BigInt(raw as number);
  }

  /**
   * Return remaining allowance as a display string (e.g. "8.5000000 USDC").
   */
  async remainingAllowanceDisplay(policyId: bigint): Promise<string> {
    const policy = await this.getPolicy(policyId);
    const remaining = await this.remainingAllowance(policyId);
    return `${fromStroops(remaining)} ${policy.asset}`;
  }

  // ─── Write Methods ──────────────────────────────────────────────────────────

  /**
   * Create a new spending policy.
   *
   * @param params  Policy creation parameters
   * @param keypair Keypair of the policy owner
   * @returns       The new policy ID
   */
  async create(params: CreatePolicyParams, keypair: Keypair): Promise<bigint> {
    const account = await this.rpc.getAccount(keypair.publicKey());

    const args = [
      nativeToScVal(params.agent, { type: 'address' }),
      nativeToScVal(toStroops(params.maxPerTx), { type: 'i128' }),
      nativeToScVal(toStroops(params.dailyLimit), { type: 'i128' }),
      nativeToScVal(params.asset, { type: 'string' }),
      nativeToScVal(params.issuer, { type: 'address' }),
      params.allowedRecipient
        ? xdr.ScVal.scvVec([nativeToScVal(params.allowedRecipient, { type: 'address' })])
        : xdr.ScVal.scvVoid(),
    ];

    const tx = new TransactionBuilder(account, {
      fee: BASE_FEE,
      networkPassphrase: this.config.networkPassphrase,
    })
      .addOperation(this.contract.call('create_policy', ...args))
      .setTimeout(30)
      .build();

    const prepared = await this.rpc.prepareTransaction(tx);
    prepared.sign(keypair);

    const sendResult = await this.rpc.sendTransaction(prepared);
    const txResult = await this._waitForConfirmation(sendResult.hash);

    // Extract policy ID from return value
    const returnVal = scValToNative(txResult);
    return BigInt(returnVal as number);
  }

  /**
   * Build an unsigned create-policy transaction for browser-wallet signing.
   *
   * @param params          Policy creation parameters
   * @param ownerPublicKey  Public key of the wallet that will sign
   * @returns               Unsigned transaction XDR (base64)
   */
  async buildCreatePolicyTx(
    params: CreatePolicyParams & { ownerPublicKey: string }
  ): Promise<string> {
    const account = await this.rpc.getAccount(params.ownerPublicKey);

    const args = [
      nativeToScVal(params.agent, { type: 'address' }),
      nativeToScVal(toStroops(params.maxPerTx), { type: 'i128' }),
      nativeToScVal(toStroops(params.dailyLimit), { type: 'i128' }),
      nativeToScVal(params.asset, { type: 'string' }),
      nativeToScVal(params.issuer, { type: 'address' }),
      params.allowedRecipient
        ? xdr.ScVal.scvVec([nativeToScVal(params.allowedRecipient, { type: 'address' })])
        : xdr.ScVal.scvVec([]),
    ];

    const tx = new TransactionBuilder(account, {
      fee: BASE_FEE,
      networkPassphrase: this.config.networkPassphrase,
    })
      .addOperation(this.contract.call('create_policy', ...args))
      .setTimeout(30)
      .build();

    const prepared = await this.rpc.prepareTransaction(tx);
    return (prepared as Transaction).toEnvelope().toXDR('base64');
  }

  /**
   * Build an unsigned revoke-policy transaction for browser-wallet signing.
   *
   * @param policyId        Policy to revoke
   * @param ownerPublicKey  Public key of the wallet that will sign
   * @returns               Unsigned transaction XDR (base64)
   */
  async buildRevokePolicyTx(
    policyId: bigint,
    ownerPublicKey: string
  ): Promise<string> {
    const account = await this.rpc.getAccount(ownerPublicKey);

    const tx = new TransactionBuilder(account, {
      fee: BASE_FEE,
      networkPassphrase: this.config.networkPassphrase,
    })
      .addOperation(
        this.contract.call('revoke_policy', nativeToScVal(policyId, { type: 'u64' }))
      )
      .setTimeout(30)
      .build();

    const prepared = await this.rpc.prepareTransaction(tx);
    return (prepared as Transaction).toEnvelope().toXDR('base64');
  }

  /**
   * Submit a signed transaction XDR (returned by Freighter) to the network.
   *
   * @param signedXdr  Base64 XDR signed by Freighter
   * @returns          Transaction hash
   */
  async submitSignedTx(signedXdr: string): Promise<string> {
    const tx = TransactionBuilder.fromXDR(
      signedXdr,
      this.config.networkPassphrase
    );
    const response = await this.rpc.sendTransaction(tx);
    await this._waitForConfirmation(response.hash);
    return response.hash;
  }

  /**
   * Execute an autonomous payment under a spending policy.
   *
   * This is the heart of Proxima. The agent signs this transaction —
   * the owner does NOT need to be present. The Soroban contract enforces
   * all spending limits on-chain.
   *
   * @param params  Payment parameters
   * @param agentKeypair  Keypair of the authorized agent
   * @returns       The payment record
   */
  async executePayment(
    params: ExecutePaymentParams,
    agentKeypair: Keypair
  ): Promise<PaymentRecord> {
    const account = await this.rpc.getAccount(agentKeypair.publicKey());

    const args = [
      nativeToScVal(params.policyId, { type: 'u64' }),
      nativeToScVal(params.recipient, { type: 'address' }),
      nativeToScVal(toStroops(params.amount), { type: 'i128' }),
      nativeToScVal(params.memo, { type: 'string' }),
    ];

    const tx = new TransactionBuilder(account, {
      fee: BASE_FEE,
      networkPassphrase: this.config.networkPassphrase,
    })
      .addOperation(this.contract.call('execute_payment', ...args))
      .setTimeout(30)
      .build();

    const prepared = await this.rpc.prepareTransaction(tx);
    prepared.sign(agentKeypair); // ← agent signs, NOT the owner

    const sendResult = await this.rpc.sendTransaction(prepared);
    const retval = await this._waitForConfirmation(sendResult.hash);

    return this._parsePaymentRecord(scValToNative(retval));
  }

  /**
   * Revoke a spending policy. Only the owner can revoke.
   */
  async revoke(policyId: bigint, ownerKeypair: Keypair): Promise<void> {
    const account = await this.rpc.getAccount(ownerKeypair.publicKey());

    const tx = new TransactionBuilder(account, {
      fee: BASE_FEE,
      networkPassphrase: this.config.networkPassphrase,
    })
      .addOperation(
        this.contract.call('revoke_policy', nativeToScVal(policyId, { type: 'u64' }))
      )
      .setTimeout(30)
      .build();

    const prepared = await this.rpc.prepareTransaction(tx);
    prepared.sign(ownerKeypair);

    const sendResult = await this.rpc.sendTransaction(prepared);
    await this._waitForConfirmation(sendResult.hash);
  }

  // ─── Helpers ────────────────────────────────────────────────────────────────

  private _buildSimulation(method: string, args: xdr.ScVal[]) {
    const source = Keypair.random();
    return new TransactionBuilder(
      { accountId: () => source.publicKey(), sequenceNumber: () => '0', incrementSequenceNumber: () => {} } as any,
      { fee: BASE_FEE, networkPassphrase: this.config.networkPassphrase }
    )
      .addOperation(this.contract.call(method, ...args))
      .setTimeout(30)
      .build();
  }

  private async _waitForConfirmation(hash: string, maxAttempts = 10): Promise<xdr.ScVal> {
    for (let i = 0; i < maxAttempts; i++) {
      await new Promise((r) => setTimeout(r, 2000));
      const status = await this.rpc.getTransaction(hash);
      if (status.status === SorobanRpc.Api.GetTransactionStatus.SUCCESS) {
        return (status as SorobanRpc.Api.GetSuccessfulTransactionResponse).returnValue!;
      }
      if (status.status === SorobanRpc.Api.GetTransactionStatus.FAILED) {
        throw new ProximaError('Transaction failed', ErrorCodes.CONTRACT_ERROR, status);
      }
    }
    throw new ProximaError('Confirmation timeout', ErrorCodes.NETWORK_ERROR);
  }

  private _parsePolicy(raw: any): SpendingPolicy {
    return {
      id: BigInt(raw.id),
      owner: raw.owner,
      agent: raw.agent,
      maxPerTx: BigInt(raw.max_per_tx),
      dailyLimit: BigInt(raw.daily_limit),
      asset: raw.asset,
      issuer: raw.issuer,
      spentToday: BigInt(raw.spent_today),
      totalSpent: BigInt(raw.total_spent),
      isActive: raw.is_active,
      createdAt: raw.created_at,
      allowedRecipient: raw.allowed_recipient ?? undefined,
      remainingAllowance: BigInt(raw.daily_limit) - BigInt(raw.spent_today),
    };
  }

  private _parsePaymentRecord(raw: any): PaymentRecord {
    return {
      policyId: BigInt(raw.policy_id),
      agent: raw.agent,
      recipient: raw.recipient,
      amount: BigInt(raw.amount),
      asset: raw.asset,
      memo: raw.memo,
      ledger: raw.ledger,
    };
  }
}
